src/vision/status_quo/alan_hates_writing_a_stream.md

```diff
@@ -0,0 +1,194 @@
+
+```rust
+async* fn sign(file: ReaderStream) -> Result<Vec<u8>, Error> {
+    let mut sig = Signature::new();
+
+    while let Some(chunk) = file.next().await? {
+        sig.push(chunk.len());
+        yield Ok(chunk)
+    }
+
+    yield Ok(sig.digest().await)
+}
+```
+
+Unfortunately, that doesn't work. The compiler first complains about the `async* fn` syntax:
+
+```
+error: expected item, found keyword `async`
+  --> src/lib.rs:21:1
+   |
+21 | async* fn sign(file: ReaderStream) -> Result<Vec<u8>, Error> {
+   | ^^^^^ expected item
+```
+
+Less hopeful, Alan tries just deleting the asterisk:
+
+```
+error[E0658]: yield syntax is experimental
+  --> src/lib.rs:27:9
+   |
+27 |         yield Ok(chunk)
+   |         ^^^^^^^^^^^^^^^
+   |
+   = note: see issue #43122 <https://github.com/rust-lang/rust/issues/43122> for more information
+```
+
+After reading about how yield is experimental, and giving up reading the 100+ comments in the [linked issue](https://github.com/rust-lang/rust/issues/43122), Alan figures he's just got to implement `Stream` manually.
+
+### Implementing `Stream`
+
+Implementing a `Stream` means writing async code in a way that doesn't _feel_ like the `async fn` that Alan has written so far. He needs to write a `poll` function and it has a lot of unfamiliar concepts:
+
+- `Pin`
+- State machines
+- `Wakers`
+
+Unsure of what the final code will look like, he starts with:
+
+```rust
+struct SigningFile;
+
+impl Stream for SigningFile {
+    type Item = Result<Vec<u8>, Error>;
+    
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context)
+        -> Poll<Self::Item>
+    {
+ 
+    }
+}
+```
+
+#### Pin :scream:
+
+First, he notices `Pin`. Alan wonders, "Why does `self` have bounds? I've only ever seen `self`, `&self`, and `&mut self` before". Curious, he reads the [`std::pin`](https://doc.rust-lang.org/std/pin/struct.Pin.html) page, and a bunch of jargon about pinning data in memory. He also reads that this is useful to guarantee that an object cannot move, and he wonders why he cares about that. The only example on the page explains how to write a ["self-referential struct"][self-ref], but notices it needs `unsafe` code, and that triggers an internal alarm in Alan: "I thought Rust was safe..."
+
+Eventually (how?), Alan realizes that the types he's depending on are `Unpin`, and so he doesn't need to worry about the unsafe stuff. It's just a more-annoying pointer type.
+
+[self-ref]: https://doc.rust-lang.org/std/pin/index.html#example-self-referential-struct
+
+#### State Machine
+
+With `Pin` hopefully ignored, Alan next notices that in the imperative style he wanted originally, he didn't need to explicitly keep track of state. The state was simply the imperative order of the function. But in a `poll` function, the state isn't saved by the compiler. Alan finds blog posts about the dark ages of Futures 0.1, when even `Future` was written with a "state machine".
```

Comment:
nit: I had to re-read the last sentence twice to get what you mean. It's not that the `Future` trait itself was defined with a state machine, but rather that futures (i.e., types that implemented `Future`) were written using state machines (and in fact still are if you manually implement `Future` for a type). Perhaps this could be worded slightly differently to avoid this hiccup in understanding?